graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,201 @@
+
+    The pagination capacity of a field is defined as the maximum number of pages we can split
+    the query results in by adding filters on this field with some confidence that the pages
+    will have similar sizes. This reasoning is local: if a filter in a different location is
+    correlated with the values on this field, the generated pages might turn out to have
+    wildly different sizes. This problem is somewhat unavoidable.
+    """
+    pagination_capacities = {}
+    for location, location_info in query_metadata.registered_locations:
+        vertex_type_name = location_info.type.name
+
+        for field_name, _ in location_info.type.fields.items():
+            key = (location.query_path, field_name)
+            if not is_meta_field(field_name):
+                if is_uuid4_type(schema_info, vertex_type_name, field_name):
+                    pagination_capacities[key] = distinct_result_set_estimates[location.query_path]
+                elif field_supports_range_reasoning(schema_info, vertex_type_name, field_name):
+                    field_value_interval = field_value_intervals.get(key, Interval(None, None))
+                    quantiles = schema_info.statistics.get_field_quantiles(
+                        vertex_type_name, field_name
+                    )
+                    if quantiles is not None:
+
+                        # Since we can't be sure the minimum observed value is the
+                        # actual minimum value, we treat values less than it as part
+                        # of the first quantile. That's why we drop the minimum and
+                        # maximum observed values from the quantile list.
+                        proper_quantiles = quantiles[1:-1]
+
+                        # Get the relevant quantiles (ones inside the field_value_interval)
+                        min_quantile = 0
+                        max_quantile = len(proper_quantiles)
+                        if field_value_interval.lower_bound is not None:
+                            min_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.lower_bound
+                            )
+                        if field_value_interval.upper_bound is not None:
+                            max_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.upper_bound
+                            )
+                        relevant_quantiles = proper_quantiles[min_quantile:max_quantile]
+
+                        # TODO take into account duplicate quantile values
+
+                        pagination_capacities[key] = min(
+                            len(relevant_quantiles) + 1,
+                            distinct_result_set_estimates[location.query_path],
+                        )
+
+    return pagination_capacities
+
+
+@dataclass
+class QueryAnalysis:
+    """A cache for analysis passes over a fixed query and fixed schema_info."""
```

Comment:
@obi1kenobi this class can simplify the codebase by a lot, and might change some APIs if fully propagated. Let me know if you see any red flags before I go on using it, since it would be a large refactor to reverse it.

Reply:
It seems fine, sure.